### PR TITLE
SONARJAVA-1735 RSPEC-3577 Test class name check

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/naming/BadTestClassNameCheck.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  private boolean hasInvalidName(@Nullable IdentifierTree className) {
+    return className != null && !pattern.matcher(className.name()).matches();
+  }
+
+  private static boolean hasTestMethod(List<Tree> members) {
+    return members.stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(tree -> ((MethodTree) tree).symbol().metadata())
+      .anyMatch(metadata -> metadata.isAnnotatedWith("org.junit.Test") || metadata.isAnnotatedWith("org.testng.annotations.Test"));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/naming/BadTestClassNameCheck.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: java-checks/src/test/java/org/sonar/java/checks/naming/BadTestClassNameCheckTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.junit.Test;
+import org.sonar.java.checks.verifier.JavaCheckVerifier;
+
+public class BadTestClassNameCheckTest {
+  @Test
+  public void test() {
+    BadTestClassNameCheck check = new BadTestClassNameCheck();
+    JavaCheckVerifier.verify("src/test/files/checks/naming/BadTestClassNameCheck.java", check);
+    // play a second time with same check to reuse pattern
+    JavaCheckVerifier.verify("src/test/files/checks/naming/BadTestClassNameCheck.java", check);
```
Comment (benzonico) [inline java-checks/src/test/java/org/sonar/java/checks/naming/BadTestClassNameCheckTest.java]: There are no explanations about why... ? I have no clue why you do play this twice.

Comment (m-g-sonar) [inline java-checks/src/test/java/org/sonar/java/checks/naming/BadTestClassNameCheckTest.java]: Only to cover the lazy initialization of the pattern ([BadTestClassNameCheck.java, L58](https://github.com/SonarSource/sonar-java/pull/890/files#diff-9084a9c491d851a91e1c344cac8337fcR58))

Comment (benzonico) [inline java-checks/src/test/java/org/sonar/java/checks/naming/BadTestClassNameCheckTest.java]: Then this should probably be part of the comment ?

**Pull request merged**